TASK: Apply rector to flow classes and tests

mficzel opened the pull request:
A rector config is added to the flow dev collection and all language level rulesets are applied up to PHP 8.0.
The Rules that caused trouble are marked to be skipped.

**Review instructions** 

The only manually edited file in this pr is the rector.php in the rest is applying rules and applying diffs from style ci.
In future changes we should make the list of skipped rules shorter.

For testing install `rector/rector` via composer and then run `rector process` in the Framework folder.

mhsdesign commented:
Noice. Early comment: rector seems to like to write exception codes in the format like `1_348_750_840` which i think is wierd ...

kitsunet commented on Neos.Flow/Classes/Validation/Validator/UniqueEntityValidator.php:
```diff
@@ -57,32 +57,32 @@ class UniqueEntityValidator extends AbstractValidator
 
-        if (count($identityProperties) === 0) {
-            throw new InvalidValidationOptionsException('The object supplied for the UniqueEntityValidator must have at least one identity property.', 1358459831);
+        if ((is_countable($identityProperties) ? count($identityProperties) : 0) === 0) {
```
it seesm not to contextualise (no surprise) but this is eg. an unnecessary addition....

kitsunet reviewed (commented):
Still haven't looked at very single file, but most of them. A lot of useful stuff but a few things that I think shouldn't happen.

albe commented on Neos.Flow/Classes/Configuration/ConfigurationManager.php:
```diff
@@ -168,15 +161,17 @@ class ConfigurationManager
      */
-    public function __construct(ApplicationContext $context)
+    public function __construct(
+        /**
```
Agree. IMO constructor properties only make sense for DTO/VOs where **ALL** properties are declared in the constructor. Having properties split between normal declaration and constructor arguments is a very bad style in my book.

albe commented on Neos.Flow/Classes/I18n/Parser/DatetimeParser.php:
```diff
@@ -78,7 +77,6 @@ public function parseDatetimeWithCustomPattern($datetimeToParse, $format, I18n\L
      * locale.
      *
      * @param string $dateToParse date to be parsed
-     * @param I18n\Locale $locale
```
I also find this weird to leave out intermediate arguments in the docblock, just because those are fully typed in code. Yes, this line does not yield additional information, but when you look over the docblock it should be a consistent view and not partial.

kitsunet replied:
agreed

mhsdesign replied:
i like this approach suggested here better: https://discuss.neos.io/t/rfc-updated-default-code-style/5836?u=marc

albe commented on Neos.Flow/Classes/Configuration/Loader/AppendLoader.php:
```diff
@@ -19,20 +19,8 @@
-    private $filePrefix;
-
-    public function __construct(YamlSource $yamlSource, string $filePrefix)
+    public function __construct(private YamlSource $yamlSource, private string $filePrefix)
```
I think in this case it's fine. All properties of the class are fully declared within the constructor.

kitsunet replied:
super and that's fine I guess and tooling will help, but going through a class hierarchy and checking for class properties and constructors like this seems super annoying. I am really not in favor in terms of readability.

albe commented on Neos.Flow/Classes/Error/DebugExceptionHandler.php:
```diff
@@ -137,7 +135,7 @@ protected function renderStatically(int $statusCode, \Throwable $exception)
             $bootstrap = Bootstrap::$staticObjectManager->get(Bootstrap::class);
             $footer .= '<tr><th>Application Context</th><td class="ExceptionProperty">' . $bootstrap->getContext() . '</td></tr>';
-            $footer .= '<tr><th>Request Handler</th><td class="ExceptionProperty">' . get_class($bootstrap->getActiveRequestHandler()) . '</td></tr>';
+            $footer .= '<tr><th>Request Handler</th><td class="ExceptionProperty">' . $bootstrap->getActiveRequestHandler()::class . '</td></tr>';
```
Looks weird but guess it's valid syntax.

kitsunet replied:
Ha I thought the same

mficzel commented on Neos.Flow/Classes/Composer/InstallerScripts.php:
```diff
@@ -76,7 +72,7 @@ public static function postPackageUpdateAndInstall(PackageEvent $event): void
         $operation = $event->getOperation();
         if (!$operation instanceof InstallOperation && !$operation instanceof UpdateOperation) {
-            throw new Exception\UnexpectedOperationException('Handling of operation of type "' . get_class($operation) . '" not supported', 1348750840);
+            throw new Exception\UnexpectedOperationException('Handling of operation of type "' . $operation::class . '" not supported', 1_348_750_840);
```
Yap i also think that this rule should be disabled

mficzel commented:
Question is would it make sense to add rector as a dev-dependency or is this a tool one uses seperately and commits the results.

mficzel committed:
TASK: Add rector config and composer script

mficzel committed:
TASK: Apply rector rules for up to php 5.6

mficzel committed:
TASK: Apply rector rules for up to php 7.0

mficzel committed:
TASK: Apply rector rules for up to php 7.1

mficzel committed:
TASK: Apply rector rules for up to php 7.4

mficzel committed:
TASK: Apply style ci adjustments

mficzel committed:
TASK: Apply rector rules for up to php 8.0

mficzel committed:
TASK: Remove rector from framework composer json

to be moved to the dest distributions

mficzel committed:
TASK: Update psalm baseline

mficzel committed:
TASK: Apply StrStartsWithRector and StrEndsWithRector

mficzel committed:
TASK: Apply StrContainsRector

and fix the issues in ActionController manually

mhsdesign commented on Neos.Cache/Classes/Backend/SimpleFileBackend.php:
```diff
@@ -261,9 +261,9 @@ public function remove(string $entryIdentifier): bool
+            } catch (\Exception) {
             }
-            usleep(rand(10, 500));
+            usleep(random_int(10, 500));
```
do we need the transformation of all the random functions?

mficzel replied:
I would say using real randomness is generally a good practice. So i would say by default yes but we might chose to use rand in places for performance reasons.

mhsdesign replied:
I can't say much about it just wanted to point it out ;)

daniellienert commented on Neos.Eel/Classes/Helper/StringHelper.php:
```diff
@@ -707,7 +707,7 @@ public function wordCount($unicodeString)
         $unicodeString = preg_replace('/[[:punct:][:digit:]]/', '', $unicodeString);
 
-        return count(preg_split('/[[:space:]]+/', $unicodeString, 0, PREG_SPLIT_NO_EMPTY));
+        return is_array(preg_split('/[[:space:]]+/', $unicodeString, 0, PREG_SPLIT_NO_EMPTY)) || preg_split('/[[:space:]]+/', $unicodeString, 0, PREG_SPLIT_NO_EMPTY) instanceof \Countable ? count(preg_split('/[[:space:]]+/', $unicodeString, 0, PREG_SPLIT_NO_EMPTY)) : 0;
```
👍

mficzel replied:
Is adjusted

daniellienert commented on Neos.FluidAdaptor/Tests/Unit/ViewHelpers/Uri/ActionViewHelperTest.php:
```diff
@@ -85,6 +85,7 @@ public function renderCorrectlyPassesAllArgumentsToUriBuilder()
      */
     public function renderThrowsViewHelperExceptionIfUriBuilderThrowsFlowException()
     {
+        $exception = null;
```
?

mhsdesign replied:
was added because of line 96 which uses $exception when its possibly undefined...

mficzel committed:
TASK: Adjust weirdness introduced by rector

While the rules had been applied correctly the code was improved by making checks obsolete.

mficzel commented on Neos.Eel/Tests/Unit/FlowQuery/FlowQueryTest.php:
```diff
@@ -455,7 +455,7 @@ public function filterCanFilterObjects($sourceObjects, $filterString, $expected)
     {
         $query = $this->createFlowQuery($sourceObjects);
-        self::assertSame(count($expectedResultArray) > 0, $query->is($filterString));
+        self::assertSame((is_countable($expectedResultArray) ? count($expectedResultArray) : 0) > 0, $query->is($filterString));
```
Yeah adjusted by adding type hints to the test method. That made the is_countable obsolete aswell.

mhsdesign commented:
@mficzel I saw that you made now some manuall changes, so we should not wait too long to merge this... (before conflicts arise)

mficzel commented:
I consider this outdated by now

mhsdesign commented:
thats the essence: (in case the branch will deleted or sth)


```php
<?php

// rector-config.php

declare(strict_types=1);

use Rector\CodeQuality\Rector\ClassMethod\OptionalParametersAfterRequiredRector;
use Rector\Config\RectorConfig;
use Rector\Php55\Rector\String_\StringClassNameToClassConstantRector;
use Rector\Php71\Rector\FuncCall\RemoveExtraParametersRector;
use Rector\Php73\Rector\FuncCall\ArrayKeyFirstLastRector;
use Rector\Php73\Rector\FuncCall\JsonThrowOnErrorRector;
use Rector\Php74\Rector\LNumber\AddLiteralSeparatorToNumberRector;
use Rector\Php74\Rector\MethodCall\ChangeReflectionTypeToStringToGetNameRector;
use Rector\Php74\Rector\Property\TypedPropertyRector;
use Rector\Php80\Rector\Class_\ClassPropertyAssignToConstructorPromotionRector;
use Rector\Php80\Rector\FuncCall\ClassOnObjectRector;
use Rector\Php80\Rector\FuncCall\TokenGetAllToObjectRector;
use Rector\Php80\Rector\FunctionLike\UnionTypesRector;
use Rector\Php80\Rector\Identical\StrEndsWithRector;
use Rector\Php80\Rector\Identical\StrStartsWithRector;
use Rector\Php80\Rector\NotIdentical\StrContainsRector;
use Rector\Php80\Rector\Switch_\ChangeSwitchToMatchRector;
use Rector\Set\ValueObject\LevelSetList;

/**
 * To execute the configured rector rules run
 * `rector process` in the Framework folder
 * after installing `rector/rector` via composer.
 *
 * For dryrun add `-n` to only show the adjustments
 */
return static function (RectorConfig $rectorConfig): void
{
    $rectorConfig->parallel();
    $rectorConfig->paths([
        __DIR__ . '/*/Classes',
        __DIR__ . '/*/Tests'
    ]);
    $rectorConfig->sets([LevelSetList::UP_TO_PHP_80]);
    $rectorConfig->skip([
        // the following rules are skipped because they caused trouble
        StringClassNameToClassConstantRector::class,
        RemoveExtraParametersRector::class,
        JsonThrowOnErrorRector::class,
        AddLiteralSeparatorToNumberRector::class,
        ChangeReflectionTypeToStringToGetNameRector::class,
        TypedPropertyRector::class,
        ArrayKeyFirstLastRector::class,
        ClassPropertyAssignToConstructorPromotionRector::class,
        UnionTypesRector::class,
        TokenGetAllToObjectRector::class,
        ClassOnObjectRector::class,
        ChangeSwitchToMatchRector::class,
        OptionalParametersAfterRequiredRector::class,

        // the autogenerated parsers have to be ignored
        __DIR__ . '/Neos.Eel/Classes/EelParser.php',
        __DIR__ . '/Neos.Eel/Classes/FlowQuery/FizzleParser.php'
    ]);
};
```

mhsdesign commented:
Hmmm i like to take another spin on this. While it seems that adjusting too much at once (even in separate commits but same pr) is not really readable i like to propose to apply at least the rules that make life easier as `str_starts_with` and the like: https://github.com/neos/flow-development-collection/pull/2706/commits/55cdb5de0d5c6219bd4001b126b977987c0f157e

What do you say? (We could also have a look at the neos codebase at one point 👀 )
